core/src/utils/forms/form-controller.ts

```diff
@@ -0,0 +1,31 @@
+/**
+ * Creates a controller that tracks whether a form control is using the legacy or modern syntax. This should be removed when the legacy form control syntax is removed.
+ *
+ * @internal
+ * @prop el: The Ionic form component to reference
+ */
+export const createLegacyFormController = (el: HTMLElement): LegacyFormController => {
+  const controlEl: HTMLElement = el;
+  let legacyControl = true;
+
+  /**
+   * Detect if developers are using the legacy form control syntax
+   * so a deprecation warning is logged. This warning can be disabled
+   * by either using the new `label` property or setting `aria-label`
+   * on the control.
+   */
+  const hasLabelProp = (controlEl as any).label !== undefined;
+  const hasAriaLabelAttribute = controlEl.hasAttribute('aria-label');
+
+  legacyControl = !hasLabelProp && !hasAriaLabelAttribute;
+
+  const hasLegacyControl = () => {
+    return legacyControl;
+  };
+
+  return { hasLegacyControl };
+};
+
+export type LegacyFormController = {
```

Comment:
I _think_ you could use `ReturnType` here, so you wouldn't need to manually maintain the type.

e.g.:
```ts
export type LegacyFormController = ReturnType<typeof createLegacyFormController>;
```

Reply:
This gave me a `Type alias 'LegacyFormController' circularly references itself.` error

Reply:
Ahh, this is because the function would then control the type, so having this:
```ts
export const createLegacyFormController = (el: HTMLIonInputElement): LegacyFormController => {
```

creates the circular reference, vs:

```ts
export const createLegacyFormController = (el: HTMLIonInputElement) => {
```

I'm not wedded to this change. Just thinking about how to avoid retyping the implementation manually.

Reply:
I'll keep the code as-is for now, but if you think of an alternative let me know.